File: web/packages/dtx-core/tests/parser.test.ts
import { describe, it, expect } from 'vitest';
import { parseDtx } from '../src/parser/parser.js';
import { Channel } from '../src/model/channel.js';

describe('parseDtx', () => {
  it('parses metadata', () => {
    const dtx = [
      '#TITLE Test Song',
      '#ARTIST Some Artist',
      '#GENRE Rock',
      '#BPM 145',
      '#DLEVEL 550',
    ].join('\n');

    const song = parseDtx(dtx);
    expect(song.title).toBe('Test Song');
    expect(song.artist).toBe('Some Artist');
    expect(song.genre).toBe('Rock');
    expect(song.baseBpm).toBe(145);
    expect(song.drumLevel).toBe(550);
  });

  it('parses WAV definitions with volume and pan', () => {
    const dtx = [
      '#WAV01 kick.wav',
      '#WAVVOL01 80',
      '#WAVPAN01 -20',
      '#WAV02 snare.wav',
      '#VOLUME02 90',
      '#PAN02 10',
    ].join('\n');

    const song = parseDtx(dtx);
    expect(song.wavTable.get(1)).toMatchObject({ path: 'kick.wav', volume: 80, pan: -20 });
    expect(song.wavTable.get(2)).toMatchObject({ path: 'snare.wav', volume: 90, pan: 10 });
  });

  it('parses BPM table', () => {
    const dtx = ['#BPM01 145', '#BPM02 90.5'].join('\n');
    const song = parseDtx(dtx);
    expect(song.bpmTable.get(1)).toBe(145);
    expect(song.bpmTable.get(2)).toBe(90.5);
  });

  it('parses chip line with four slots of snare', () => {
    // Snare (0x12) at four equal positions (each 1/4 of the measure).
    const dtx = ['#BPM 120', '#WAV01 s.wav', '#00012:01010101'].join('\n');
    const song = parseDtx(dtx);

    const snare = song.chips.filter((c) => c.channel === Channel.Snare);
    expect(snare).toHaveLength(4);
    expect(snare.map((c) => c.tick)).toEqual([0, 96, 192, 288]);
    expect(snare.every((c) => c.wavId === 1)).toBe(true);
  });

  it('skips 00 slots', () => {
    const dtx = ['#00013:00010002'].join('\n');
    const song = parseDtx(dtx);
    const bd = song.chips.filter((c) => c.channel === Channel.BassDrum);
    expect(bd).toHaveLength(2);
    expect(bd.map((c) => c.tick).sort((a, b) => a - b)).toEqual([96, 288]);
    expect(bd.map((c) => c.wavId).sort()).toEqual([1, 2]);
  });

  it('parses BPMChangeExtended chips (channel 0x08)', () => {
    const dtx = ['#BPM 120', '#BPM01 180', '#00108:01000000'].join('\n');
    const song = parseDtx(dtx);
    const bpmChips = song.chips.filter((c) => c.channel === Channel.BPMChangeExtended);
    expect(bpmChips).toHaveLength(1);
    expect(bpmChips[0]?.bpmId).toBe(1);
    expect(bpmChips[0]?.measure).toBe(1);
  });

  it('ignores comments and blank lines', () => {
    const dtx = [
      '; a comment',
      '',
      '#TITLE Foo   ; inline',
      '  ',
    ].join('\n');
    const song = parseDtx(dtx);
    expect(song.title).toBe('Foo');
  });

  it('strips UTF-8 BOM', () => {
    const dtx = '\uFEFF#TITLE BomSong';
    const song = parseDtx(dtx);
    expect(song.title).toBe('BomSong');
  });
});

File: native/quest-spatial/app/src/test/kotlin/com/dtxmania/quest/dtxcore/parser/ParserTest.kt
package com.dtxmania.quest.dtxcore.parser

import com.dtxmania.quest.dtxcore.model.Channel
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

/**
 * Ported 1:1 from `web/packages/dtx-core/tests/parser.test.ts`.
 */
class ParserTest {
    @Test fun `parses metadata`() {
        val dtx = listOf(
            "#TITLE Test Song",
            "#ARTIST Some Artist",
            "#GENRE Rock",
            "#BPM 145",
            "#DLEVEL 550",
        ).joinToString("\n")

        val song = parseDtx(dtx)
        assertEquals("Test Song", song.title)
        assertEquals("Some Artist", song.artist)
        assertEquals("Rock", song.genre)
        assertEquals(145.0, song.baseBpm, 0.0)
        assertEquals(550, song.drumLevel)
    }

    @Test fun `parses WAV definitions with volume and pan`() {
        val dtx = listOf(
            "#WAV01 kick.wav",
            "#WAVVOL01 80",
            "#WAVPAN01 -20",
            "#WAV02 snare.wav",
            "#VOLUME02 90",
            "#PAN02 10",
        ).joinToString("\n")

        val song = parseDtx(dtx)
        val w1 = song.wavTable[1]!!
        assertEquals("kick.wav", w1.path)
        assertEquals(80, w1.volume)
        assertEquals(-20, w1.pan)
        val w2 = song.wavTable[2]!!
        assertEquals("snare.wav", w2.path)
        assertEquals(90, w2.volume)
        assertEquals(10, w2.pan)
    }

    @Test fun `parses BPM table`() {
        val dtx = listOf("#BPM01 145", "#BPM02 90.5").joinToString("\n")
        val song = parseDtx(dtx)
        assertEquals(145.0, song.bpmTable[1]!!, 0.0)
        assertEquals(90.5, song.bpmTable[2]!!, 0.0)
    }

    @Test fun `parses chip line with four slots of snare`() {
        // Snare (0x12) at four equal positions (each 1/4 of the measure).
        val dtx = listOf("#BPM 120", "#WAV01 s.wav", "#00012:01010101").joinToString("\n")
        val song = parseDtx(dtx)

        val snare = song.chips.filter { it.channel == Channel.SNARE }
        assertEquals(4, snare.size)
        assertEquals(listOf(0, 96, 192, 288), snare.map { it.tick })
        assertTrue(snare.all { it.wavId == 1 })
    }

    @Test fun `skips 00 slots`() {
        val dtx = "#00013:00010002"
        val song = parseDtx(dtx)
        val bd = song.chips.filter { it.channel == Channel.BASS_DRUM }
        assertEquals(2, bd.size)
        assertEquals(listOf(96, 288), bd.map { it.tick }.sorted())
        assertEquals(listOf(1, 2), bd.mapNotNull { it.wavId }.sorted())
    }

    @Test fun `parses BPMChangeExtended chips (channel 0x08)`() {
        val dtx = listOf("#BPM 120", "#BPM01 180", "#00108:01000000").joinToString("\n")
        val song = parseDtx(dtx)
        val bpmChips = song.chips.filter { it.channel == Channel.BPM_CHANGE_EXTENDED }
        assertEquals(1, bpmChips.size)
        assertEquals(1, bpmChips[0].bpmId)
        assertEquals(1, bpmChips[0].measure)
    }

    @Test fun `ignores comments and blank lines`() {
        val dtx = listOf(
            "; a comment",
            "",
            "#TITLE Foo   ; inline",
            "  ",
        ).joinToString("\n")
        val song = parseDtx(dtx)
        assertEquals("Foo", song.title)
    }

    @Test fun `strips UTF-8 BOM`() {
        val dtx = "\uFEFF#TITLE BomSong"
        val song = parseDtx(dtx)
        assertEquals("BomSong", song.title)
    }
}
